RFC: Mark all safety relevant files with a SPDX tag

romkell opened the pull request:
# PR closed in favour of issue #49829

This PR is closed in favour of issue #49829.
Please continue the discussions there.

# RFC: Mark all safety scope relevant files with a SPDX tag

## Concerning this PR and the 800+ files changed
1. **Please do not start with reviewing any of the files yet!** It is a rough assignment judged by the filenames and folder the files are in based on a Zepyhr OS architecture block diagram with some blocks marked as in the safety scope from page 18 of [Zephyr-Overview-2022Q1-Public.pdf](https://zephyrproject.org/wp-content/uploads/sites/38/2022/02/Zephyr-Overview-2022Q1-Public.pdf). **There will be lots of false positives and negatives right now.**
1. The idea is to raise a discussion on the methods how to manage what of the Zephyr OS eco system (source tree) is in the safety scope (what will undergo safety certification) and what not.
1. The second idea behind was to get a feeling how many files will be affects
1. I now split it into 3 commits
   1. The safety related sources (basically the C code, still some 500+)
   1. The T3 tools and related files (some 250+)
   1. The T2 tools and related files (some 130+)

## Safety application development implications in short (and simplified)
This section provides a very brief overview of the implications of developing a (SIL3) safety application according to IEC 61508.
- IEC61508 defines four safety integrity levels: SIL1, SIL2, SIL3 and SIL4
  - Safe applications may fail, but in a safe manner (at the cost of availability)
  - The higher the SIL number the lower the probability of a uncontrolled dangerous failure is accepted [1/h]
    - SIL1: ≥ 10 E–6 to < 10 E–5
    - SIL2: ≥ 10 E–7 to < 10 E–6
    - SIL3: ≥ 10 E–8 to < 10 E–7
    - SIL4: ≥ 10 E–9 to < 10 E–8
- That applies to hardware and software, for the software aspect:
- All code needs to be written against a specification on typically three levels
  - System / software requirements specification (**SysRS** or **SRS**)
  - Software architecture and interface specification (**SAIS**)
  - Software module  design specification (**SMDS**)
- All **test code** needs to be written against a test specification on typically three levels
  - System / software **test** specification (STS) - securing that the **SRS** was implemented 
  - Software integration **test** specification (SITS) - securing that the **SAIS** was implemented
  - Software module **test** specification (SMTS) - securing that the **SMDS** was implemented
- All code involved needs to be tested with
  - **System/Software Tests** - showing that the **System / software specification** is fulfilled - not more, not less
  - **Integration Tests** - showing that the **Software integration specification** is fulfilled
  - **Module (Unit) Tests** - showing that the **Software module specification** is fulfilled
    - that 100% entry point coverage was reached (SIL1, 2, 3, 4)
    - that 100% statement coverage was reached (SIL2, 3, 4)
    - that 100% branch coverage was reached (SIL3, 4)
    - that 100% condition (MC/DC) coverage was reached (SIL4)
    -  (exceptions to be justified, e.g. covered by code review as substitution)
- Traceability (linkage) must be given from the top level specifications down into (test) code: V-Model
   - SRS <------------- verifies ------------- STS
      - SAIS <-------- verifies --------- SITS
         - SMDS <--- verifies --- SMTS
            - ----- Code -----

   - SRS <-- refines -- SAIS <-- refines -- SMDS 
   - STS <-- refines -- SITS <-- refines -- SMTS
   - STS <-- implements -- (Software test code)
   - SITS <-- implements -- (Software integration test code) 
   - SMTS <-- implements -- (Software Module test code)

- Tools used must be qualified for safety development
   - Three categories exist as defined by ISO 61508-4 3.2.11: T1, T2 and T3
      - **T1**: generates no outputs which can directly or indirectly contribute to the executable code
(e.g. text editor, design describing tool, ...) - **No qualification required**
      - **T2**: errors in the tool can fail to reveal defects but cannot directly create errors in the executable software - **Qualification required**
(e.g. test generator, test coverage tool, static analysis tool, ...)
      - **T3**: can directly or indirectly contribute to the executable code - **Qualification required**
(e.g. code generator, optimizing compiler, linker, ...)

- Non-safety applications running on the same system may not influence the safety application in any way - evidence must be given that this is the case; safe and non-safe applications on the same MCU without MPU/MMU are basically impossible and even with, had to prove that the MPU/MMU guarantees a fail-safe reaction!
- Therefore and as consequence of the above, mixing safe and non-safe code, in a eco system / framework where both is present, must be prevented - evidence must be given
- Zephyr OS as a framework can, strictly spoken, not be certified, it is the applications (HW and SW) build on and using Zephyr OS that will be certified.
- Zephyr OS can be, lets call it pre-certified, so that someone who builds a safe application using Zephyr OS does not have to prove safety fitness of source code and tools anymore but can treat it similar to a safety certified library with all the papers (safety collaterals) available.

## "Why tagging files?" or "The Use Cases"

- The project aims to certify the core part of the code as IEC 61508 SIL3
- Having a clear indication in the code base what is safety relevant and what not
  - provides freedom to the developer on the non-safety side and
  - creates awareness to the developer on the safety side at the beginning of a development not on PR review (avoiding hard to impossible to certify implementations)
  - creates the security for the certifying assessor/body (TÜV/exida) that what is non-safe and safe is distinguishable
- Allows to create a (tool supported) process to involve the safety architect early.
- Allows to create automated configurations for tools treating the safety and non-safety side differently (e.g. static code analyzer)
- Enables a safety user using the spdx tooling to verify and prove to TÜV/exida that only safety certified code was used
- Marking the source files will very likely outdate less than a "external database" e.g. yaml file, since it is part of the PR review process, updating an external database gets forgotten easily

## Method
All files being in the safety scope which will undergo TÜV/exida certification shall be marked with a
- "SPDX-FileComment: IEC-61508-SIL3" for Zephyr OS source code
- "SPDX-FileComment: IEC-61508-T2" for T2 class tools
- "SPDX-FileComment: IEC-61508-T3" for T3 class tools
tag.

## Tag handing "Auditable branches" vs "main"
I see two or three options how to work

1. **Do it on both immediately**
   1. Advantages
       1. Have the opporunity to handle most use cases as above immediatey
       1. Avoid safety contradicting developments in "main"
       1. Have "main" closer to "safety ready" on tagging next LTS and auditable
   1. Disadvantages
       1. Safety Committee will have to take care about both
       1. "main" developer will get confronted with safety topics
1. **Only do safety on "Auditable Branches" and never on "main"**
   1. Advantages
       1. Safety Committee can exclusively focus on auditables
       1. "main" developer do not need to care about safety topics
   1. Disadvantages
       1. Over the "main" releases, there is a gowing gap between the last and the next "auditable" branch - this gap will have to be closed (cleaned up) every time again
       1. There is hardly any control over "safety un-friendly" implementation in the safety scope
1. **Do it on "1. Auditable Branch" first, later, when finished forward merge to "Main"**
   1. Advantages and Disadvantages are at first those of 2. and later of 1.

Which way the project decides to go:
Unless the "auditable only" path is choosen, not all safety relevant artifacts (*) need to be treated equally concerning introduction on "main".
- Safety handling in the code is critical
- Updating test implementation is do-able, but since code too, it might be a lot of effort and is therefore rather critical
- Updating requirement from one "auditable" to a next is do-able, but unnessecary effort when it could be maintained on "main"

(*) all inputs required to do safety, not the outputs of the safety certification nor the safety plans

### Presumable impact to maintainer
First of all, I would assume that the code in the safety scope generally is mature code, which does not have to be turned upside down any more (otherwise the module tests will have to be rewritten too). 

- When changing a IEC-61508-xxx marked file, consult the safety architect early
- When splitting such files, copy the IEC-61508-xxx SPDX tag to the new files as you probaly would also do with the license SPDX tag
- On PR merge, assign the safety architect as reviewer

## Thoughts and suggestions
- Use a more generic prefix, Safety- instead of IEC-61508- or will we add tags ISO26262-, EN50128- DO178- for the more specific standard once we would certify them?

## Annex

(https://www.tuvsud.com/en/-/media/global/pdf-files/webinar-presentations/tuev-sued_webinar_2018-01-16_finding-the-right-software-tools-for-functional-safety-projects.pdf)

aescolar commented on subsys/logging/log_backend_native_posix.c:
```diff
@@ -3,6 +3,7 @@
  * Copyright (c) 2018 Intel Corporation
  *
  * SPDX-License-Identifier: Apache-2.0
+ * SPDX-FileComment: IEC-61508-SIL3
```
Tagging native_posix (or POSIX arch related) files would seem a mistake. (this is a test platform, so none of this code should end up in any embedded binary)

romkell replied:
Very likely mistakenly marked.

romkell replied:
Please see 

## Concerning this PR and the 800+ files changed

in the description.

aescolar commented on include/drivers/console/native_posix_console.h:
```diff
@@ -2,6 +2,7 @@
  * Copyright (c) 2018 Oticon A/S
  *
  * SPDX-License-Identifier: Apache-2.0
+ * SPDX-FileComment: IEC-61508-SIL3
```
Also seems like a false positive

romkell replied:
Very likely mistakenly marked.

romkell replied:
Please see 

## Concerning this PR and the 800+ files changed

in the description.

carlocaione commented:
> Files which have a such tag serve as indication to the developer
that they are potentially changing safety relevant code parts (including
tools). This may serve as trigger to involve the TSC / Safety Committee
and gives the freedom to a less restrictive change process in the
remaining code base.

This looks unrealistic to say the least given that all the files are now marked.

andyross commented:
Seems like this needs to be done in tandem with some level of automated maintenance software, no?  It's all great to label files statically in the tree (though others are pointing out that this is a bit too aggressive), but we add files and chop them up all the time.  And most developers (well, me, at least) just aren't going to be able to remember the rules for what needs to be labelled and why, so these tags are going to rot pretty rapidly.

Maybe have some way (parse zephyr.lst?) to automatically detect if some "canary safety build" or something includes code  from untagged files?

romkell commented:
> Seems like this needs to be done in tandem with some level of automated maintenance software, no? It's all great to label files statically in the tree (though others are pointing out that this is a bit too aggressive), but we add files and chop them up all the time. And most developers (well, me, at least) just aren't going to be able to remember the rules for what needs to be labelled and why, so these tags are going to rot pretty rapidly.
> 
> Maybe have some way (parse zephyr.lst?) to automatically detect if some "canary safety build" or something includes code from untagged files?

I see your point of chopping files and and possibly not knowing what part was considered safe and should be marked again.
I understand that most developers main interest is not safety and consider it too aggressive. On the other hand in an eco-system such as Zephyr OS certification authorities want to see all the certified code to be tested, documented etc. We need to have a mechanism to indicate what is in and what out.

What are other options:
- parsing zephyr.lst is fine but to verify it again what?
  - some kind of database telling what is in the safety scope and what not? Who would maintain that? That database would rot even faster, I guess, because it is detached from the code completely.
  - Use KConfig #ifdef CONFIG_SAFETY or similar
  - Mark all safety related functions with /* @in_safety_scope */ or similar so those information can be copied when splitting files. Quite some micro management, but easier to handle for those moving these things around

Another option would be to leave "main" untouched from such mechanisms and only do and re-do it on each x.y-auditable branch which will be certified. But how then to establish a process detecting "safety dangerous" implementations on main early such as dynamic stack allocations done by Ch. Friedt (luckily he was aware of it).

mbolivar-nordic commented on include/devicetree/clocks.h:
```diff
@@ -7,6 +7,7 @@
  * Copyright (c) 2020, Linaro Ltd.
  *
  * SPDX-License-Identifier: Apache-2.0
+ * SPDX-FileComment: IEC-61508-SIL3
```
Provided that you can document the impact of this comment on a macro-only API like devicetree.h in a way that I can understand and enforce, I have no fundamental objections to this. I will need training, though.

However, have you considered the impact on the generated devicetree_unfixed.h header itself?

romkell replied:
The question here is if these kind of SPDX tags remain on the LTS vX.Y-auditable and we repeat that for each of the v(X+n).Z-auditable's from auditable to auditable, or if such a mechanism should be merged to main.

The idea of bringing that into main and into the source files are:
- maintaining such information in a "database" separate from the source files will out-date faster than when in the files
- It also indicates in which code parts special care has to be taken, not to implement features and mechanisms which are hard to impossible to certify for safety. -> trigger the safety architect

Freedom to implement anything anyhow and safety are, to my experience, somewhat contradicting.

In the end there might be a discussion required how far the community is willing to support safety on main or if safety always is a "auditable" thing only, the safety guys having to cope with what they get from each LTS.

mbolivar-nordic replied:
> The question here is if these kind of SPDX tags remain on the LTS vX.Y-auditable and we repeat that for each of the v(X+n).Z-auditable's from auditable to auditable, or if such a mechanism should be merged to main.

I can see how this is the important question for you. For me as a DT maintainer, the important question is if you can specify:

>  the impact of this comment on a macro-only API like devicetree.h in a way that I can understand and enforce

Does that make sense?

romkell replied:
As I wrote in the description intro (only lately), I did the marking very roughly on suspicion of the block diagram (link in description). Whether or not exactly this file should get a tag or not I do not have sufficient knowledge in the different area to judge in detail. I depend on the people that maintain those files to finally understand what the purpose of the files are and if they should get a tag or not. I will happily remove any file not required - the less the better - but we will still need to get all required.

The impact for such marked files will be:
- for Zephyr OS source code is change
  - the requirements specifications will have to be updated
  - the test specification will have to be updated
  - the module / unit test will have to be update
- for Zephyr OS tools such as dt generator, ztest, etc.
  - tool qualification needs to be repeated in next auditable
    - updating all paper work which is similar to the Zephyr OS source code but very likely not in that extend

If the said above frightens the developers to change anything, there will be no further progress in what is considered the safety scope.
If we want to do repeated pre-certified auditable tags from LTS-es we will have to accept that we need do that work.
To me the point is more, if we are able to make the changes to the safety scope a controlled process or will be surprised with what has changed when the next auditable is branched.

I guess the main impact will be that the safety architect gets an early chance to get involved and have an eye on it and once we have requirements and (safety-)tests in place that all that gets updated and does not out-date.

mbolivar-nordic commented on include/drivers/led_strip.h:
```diff
@@ -2,6 +2,7 @@
  * Copyright (c) 2017 Linaro Limited
  *
  * SPDX-License-Identifier: Apache-2.0
+ * SPDX-FileComment: IEC-61508-SIL3
```
Do you really need to safety certify this API and its drivers? This seems like overkill. I am opposed to this.

romkell replied:
The driver model was marked as in the safety scope in the pdf (see description) which I interpret as the API. Where as the drivers implementation since very often MCU specific is not (yet) and has to be done by the Zephyr OS safety application user.

Changing the driver API will very likely cause the drivers to be changed and re-certified (even though not by the project) by a safety user of Zephyr OS. Therefore I can understand why the driver model was marked as safety scope.
If all drivers and which driver, is subject to discussions and definitions.

mbolivar-nordic commented on scripts/dts/gen_defines.py:
```diff
@@ -3,6 +3,7 @@
 # Copyright (c) 2019 - 2020 Nordic Semiconductor ASA
 # Copyright (c) 2019 Linaro Limited
 # SPDX-License-Identifier: BSD-3-Clause
+# SPDX-FileComment: IEC-61508-SIL3
```
Although I can see the issue here (this file generates something that basically every application uses), I don' believe I have ever seen any proposals related to the process of safety certification for Python files (or CMake files). I have only read such for C files. What am I missing?

romkell replied:
I (now) tried to explain in the description concerning the IEC61508 tool classification.
Since gen_defines.py is a code generator (I assume) it is a T3 tool and requires qualification. In a later commit the SPDX-FileComment: reads IEC-61508-T3.

romkell committed:
Mark all safety relevant file with a SPDX tag

All file being in the safety scope which will undergo TÜV/exida
certification are marked with a "SPDX-FileComment: IEC-61508-SIL3" tag.

1. When building an application and running the spdx tool over the
build folder, a safe application may only comprise such marked files.
This is a mean to allow users of the Zepyhr OS ecosystem separate
safe from non-safe application in a documented manner as preparation
for their TÜV/exida certification.

2. Files which have a such tag serve as indication to the developer
that they are potentially changing safety relevant code parts (including
tools). This may serve as trigger to involve the TSC / Safety Committee
and gives the freedom to a less restrictive change process in the
remaining code base.

Signed-off-by: Roman Kellner <rkellner@baumer.com>

romkell committed:
Mark all T3 tool files with a corresp. SPDX tag

Tools generating output which directly or indirectly may
contribute (including data) to the safety related application binary.
All files belonging to such a tool are marked with the SPDX tag
IEC-61508-T3.

Signed-off-by: Roman Kellner <rkellner@baumer.com>

romkell committed:
Mark all T2 tool files with a corresp. SPDX tag

Tools which support the verification of the design or executable code,
where deviations in the tool may lead to failure of recognizing errors
but may not introduce any deviation in the safety related software.
All files belonging to such a tool area marked with the SPDX tag
IEC-61508-T2.

Signed-off-by: Roman Kellner <rkellner@baumer.com>

gregshue commented:
In general, this proposal makes very good sense.  The different sub-topics should be addressed individually:

> "SPDX-FileComment: IEC-61508-SIL3" for Zephyr OS source code
We need to recognize this tag is appropriate to apply outside the Zephyr OS source code and will eventually need to appear in HAL modules.

> the drivers implementation since very often MCU specific is not (yet) and has to be done by the Zephyr OS safety application user.
Is there something preventing the hardware vendor HAL contributor from generating the same artifacts for the safety application user (i.e., end product developer)?

> Suggestions: Use a more generic prefix, ...
I prefer seeing the explicit standard tags to something more generic. 
- As an integrator, it most clearly indicates what level of certification each file is intended to support.  
- As a contributor, it most clearly indicates the level of rigor that I must apply to this file.  
- As a downstream module developer, it helps me distinguish which tools and processes I may reuse for my own modules and executables.
- As a participant in the Security WG, it establishes a reasonable pattern for indicating the set of security standards to be achieved.  This is particularly important when looking to integrate multiple downstream modules where each was designed to meet a different security standard.

> Maybe have some way (parse zephyr.lst?) to automatically detect if some "canary safety build" or something includes code from untagged files?

Whatever solution we design must be long-lived to be practical.  This means:
1. It must consider what happens when the auditable scope grows (planned).
2. It needs to be applicable to downstream modules which may include new CPU architectures, drivers, etc.
3. It needs to coexist in source code reused (same file, same branch) with other build systems and ecosystems.

> I understand that most developers main interest is not safety
Agreed.  I expect most developers main interest is security.  It will be interesting to see what process security requires.

> Another option would be to leave "main" untouched from such mechanisms and only do and re-do it on each x.y-auditable branch which will be certified.
Applying this strategy for security changes limits secure FW to only be released against LTS branches.  That notably impacts the ability to migrate a secure device design to newer hardware in response to chip availability and/or new technologies. It seems the same rationale applies to safety as well.

nashif commented:
> > "SPDX-FileComment: IEC-61508-SIL3" for Zephyr OS source code
> 
> We need to recognize this tag is appropriate to apply outside the Zephyr OS source code and will eventually need to appear in HAL modules.

why?

gregshue commented:
> why?

1. We can't stop anyone from using this.
2. Downstream users will create other modules with drivers that need to be included in safety-critical applications, and will try to reuse the Zephyr-provided mechanisms to warn if non-tagged files are included.
3. The auditable branch scope is already planned to grow larger in successive LTS.
4. The zephyr-repo native HAL implementation files for emulated systems will need to be tagged in order to generate a safety-critical executable.  An analogous tagging should be expected in HAL content provided in other modules (whether in zephyrproject-rtos or downstream).
5. This tag is essentially a commitment that this file and associated tests will be kept in compliance with IEC-61508-SIL3 certifiability.  That same commitment is appropriate (if not necessary) for other code in a safety-critical executable.

mbolivar-nordic commented:
> We need to

@gregshue I believe I have seen you begin several sentences this way.

I think it is likely to cause confusion for people when you speak this way. It may falsely give people the impression that you are a maintainer or collaborator in the zephyr project, and that you are speaking in that capacity.

Since you are neither, I would like to ask that you stop speaking in this way unless you are in fact using "we" to mean some other organization where you have the authority to dictate needs.

nashif commented:
> 1. We can't stop anyone from using this.

You can't stop anyone from doing anything, we are talking about the zephyr project and the code we control and have the capacity to change and manage. If someone decided to follow the zephyr's project model for doing this or that, be my guest, but lets stay focused on the what the zephyr project does and control.

> 2\. Downstream users will create other modules with drivers that need to be included in safety-critical applications, and will try to reuse the Zephyr-provided mechanisms to warn if non-tagged files are included.

Up to you to track your scope any way you want.

> 3\. The auditable branch scope is already planned to grow larger in successive LTS.

yes, maybe, but it will not grow into HALs and modules we do not control. And, lets do baby steps and focus on what this RFC is about, it is not about modules or HALs, this is about zephyr content.

> The zephyr-repo native HAL implementation...

what is that? and why would a HAL be tagged?

> 5\. This tag is essentially a commitment that this file and associated tests will be kept in compliance with IEC-61508-SIL3 certifiability.

If this is a "committment" I will be the first to stand against this. This is meta-data for accounting and tooling, THERE IS NO COMMITMENT, there should not be any hints of commitments.

gregshue commented:
> Since you are neither [a maintainer or collaborator],

Last I knew I am a collaborator on the `example-application` repository.

mbolivar-nordic commented:
> Last I knew I am a collaborator on the `example-application` repository.

```
$ grep gregshue MAINTAINERS.yml
$
```

gregshue commented:
I did not speak from the perspective of a collaborator within the Zephyr Project, but as a Zephyr Project contributor about my needs as a collaborator on a downstream, out-of-tree Zephyr module and as an end user potentially making IoT devices needing safety certification.

nashif commented:
> a collaborator on a downstream, out-of-tree Zephyr module

and which one is that?

gregshue commented:
I am a collaborator on https://github.com/bacnet-stack/bacnet-stack.

I also am also in the process of contributing to https://github.com/mcu-tools/mcuboot.

gregshue commented:
> we are talking about the zephyr project and the code we control and have the capacity to change and manage. 

If that was the limits of what the Zephyr Project is concerned with then there is no reason to be worried about breaking out-of-tree/downstream users.

> You can't stop anyone from doing anything,

>  it will not grow into HALs and modules we do not control.

These are contradictory statements.

> this is about zephyr content.

This is about how to identify safety-certifiable content in a way that can also apply for identifying security-certifiable content.  If the **mechanism** isn't suitable outside the zephyr repository, then it doesn't meet the needs of some end users.

nashif commented:
> > we are talking about the zephyr project and the code we control and have the capacity to change and manage.
> 
> If that was the limits of what the Zephyr Project is concerned with then there is no reason to be worried about breaking out-of-tree/downstream users.
> 
> > You can't stop anyone from doing anything,
> 
> > it will not grow into HALs and modules we do not control.
> 
> These are contradictory statements.
> 
> > this is about zephyr content.
> 
> This is about how to identify safety-certifiable content in a way that can also apply for identifying security-certifiable content. If the **mechanism** isn't suitable outside the zephyr repository, then it doesn't meet the needs of some end users.

This is going nowhere. I am out.

romkell commented:
> > We need to
> 
> @gregshue I believe I have seen you begin several sentences this way.
> 
> I think it is likely to cause confusion for people when you speak this way. It may falsely give people the impression that you are a maintainer or collaborator in the zephyr project, and that you are speaking in that capacity.

I understood that as "we" "the community" in what ever role.

**Concerning out-of-tree code:**

@gregshue takes a Zepyhr OS eco system end user position, the same position as Baumer is in. When building a safe application, as a end user I would want that the west SPDX output tells me, that I have used only certified code. Zephyr OS applications without peripherals drivers  and a HAL are limited to the host emulation or qemu (did I forget something?).
Hence as a end user I need to be able to create safe applications running on a board including HAL.
From that perspective and that I would like to get a report out of SPDX that I only used certified artifact, out-of-tree sources would have to be included and to carry that tag too.

But it is not in the control of the project but of the HAL vendor.

As long as this is a Zephyr OS proprietary solution, it will be difficult that it spreads elsewhere.
I understood @kestewart that she takes the approach to the SPDX standardization board (what ever the organizations name is) and tries to create an official solution (e.g. SPDX-Certification) instead of (mis-)using the SPDX-FileComment tag for that. But that might take a while to become officially released.
But It is a perspective. Maybe @kestewart can comment on that.

Finally, it is the out-of-tree source code providers and maintainers decision, whether or not he uses that mechanism. No way to force them. The more official, the more the likelihood that it will be adopted.

Using it for the inside-tree code is a start. Getting a SPDX report on that for the inside-tree is a start and already helpful to the end user. 
Get a full coverage report for all files compiled in is deluxe, but might become true in the future.

**Concerning commitment:**

I feel that there should be some kind of commitment connected. Otherwise it is just useless tags.
- As an example for safety
  1. developers /maintainers take special care that changes are done in a manner, that do not hinder the respective re-certification.
  2. If developers / maintainers, do not know what the implications are, that advice is seeked from a dedicated role (safety architect, security architect, some committee etc.) or/and documentation (safety do's and dont's)
  3. That the roles listed above are invited to review and are given the authority to ask for a changed implementation (once the kernel is full of mallocs and frees where you can hardly prove that there is no constellation where you run out of heap, safety certification is history)
  4. That certain to be defined special processes are followed for that code (as indicated above with reviewer invitation)

I would appreciate that further conversations continue in the issue #49829

gregshue commented:
@romkell Very well said.  I like the idea of continuing this discussion in #49829.  Please cross-post the main body of your comment there (with a link back here) so that we have good context for responses.
